lua/bufferline/get-icon.lua

```diff
@@ -26,7 +26,20 @@ local function get_icon(buffer_name, filetype)
     extension = vim.fn.matchstr(basename, [[\v\.@<=\w+$]], '', '')
   end
 
-  return web.get_icon(basename, extension, { default = true })
+  local iconChar, iconHl = web.get_icon(basename, extension, { default = true })
+
+  if iconHl and vim.fn.hlexists(iconHl..buffer_status) < 1 then
+    local function get_attr(group, attr)
+      return vim.fn.synIDattr(vim.fn.synIDtrans(vim.fn.hlID(group)), attr)
+    end
```

Comment:
I'm don't know how luajit works, but unless it really needs to be a closure I'd move `get_attr` outside of this as to make sure luajit doesn't create a new instance of that function each time `get_icon` is called.

Also the body of the function could use `nvim_get_hl_by_name().{foreground,background}` with `printf` as in `highlights.vim`. Less vim functions called means less overhead, and it's also more readable. `nvim_get_hl_by_name` is only available in 0.5 I think but it's ok because barbar already requires it.

Reply:
LuaJIT will not needlessly gc this closure during intensive calculations, see [here](https://www.lua.org/gems/sample.pdf) for more (bottom of page 18). I'll break it out anyway, just in case.

___

`nvim_get_hl_by_name()` function doesn't quite work in that way. Here's an example to show what I mean:

```vim
echo nvim_get_hl_by_name('BufferInactive', v:true).background
" Output: 5263440

echo synIDattr(synIDtrans(hlID('BufferInactive')), 'bg#')
" Output:  #505050
```

This is because `nvim_get_hl_by_name` reports as `RGB`, which can't be used in a highlight call. Further, `nvim_set_hl()` only sets highlights with regard to a namespace (which isn't what we want).

I actually tried `nvim_get_hl_by_name()` first, but then later switched to `synIDattr()`.